torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,259 @@
+                (defaults to output dimension of model).
+            output_dim: Number of output dimensions in projection head
+                (defaults to output dimension of model).
+            lr: Learning rate (0.3 x batch_size / 256 is recommended).
+            weight_decay: Weight decay coefficient (1e-6 for v1 or 1e-4 for v2).
+            temperature: Temperature used in NT-Xent loss.
+            memory_bank_size: Size of memory bank (0 for v1 or 64K for v2).
+            gather_distributed: Gather negatives from all GPUs during distributed
+                training (ignored if memory_bank_size > 0).
+            augmentations: Data augmentation.
+
+        Raises:
+            AssertionError: If an invalid version of SimCLR is requested.
+
+        Warns:
+            UserWarning: If hyperparameters do not match SimCLR version requested.
+        """
+        super().__init__()
+
+        # Validate hyperparameters
+        assert version in range(1, 3)
+        if version == 1:
+            if layers > 2:
+                warnings.warn("SimCLR v1 only uses 2 layers in its projection head")
+            if memory_bank_size > 0:
+                warnings.warn("SimCLR v1 does not use a memory bank")
+        elif version == 2:
+            if layers == 2:
+                warnings.warn("SimCLR v2 uses 3+ layers in its projection head")
+            if memory_bank_size == 0:
+                warnings.warn("SimCLR v2 uses a memory bank")
+
+        self.save_hyperparameters(ignore=["augmentations"])
+        self.augmentations = augmentations
+
+        # Create backbone
+        self.backbone = timm.create_model(
+            model, in_chans=in_channels, num_classes=0, pretrained=weights is True
+        )
+
+        # Load weights
+        if weights and weights is not True:
+            if isinstance(weights, WeightsEnum):
+                state_dict = weights.get_state_dict(progress=True)
+            elif os.path.exists(weights):
+                _, state_dict = utils.extract_backbone(weights)
+            else:
+                state_dict = get_weight(weights).get_state_dict(progress=True)
+            self.backbone = utils.load_state_dict(self.backbone, state_dict)
+
+        # Create projection head
+        input_dim = self.backbone.num_features
+        if hidden_dim is None:
+            hidden_dim = input_dim
+        if output_dim is None:
+            output_dim = input_dim
+
+        self.projection_head = SimCLRProjectionHead(
+            input_dim, hidden_dim, output_dim, layers
+        )
+
+        # Define loss function
+        self.criterion = NTXentLoss(temperature, memory_bank_size, gather_distributed)
+
+        # TODO
+        # v1+: add global batch norm
+        # v2: add selective kernels, channel-wise attention mechanism
+
+    def forward(self, batch: Tensor) -> Tensor:
+        """Forward pass of the model.
+
+        Args:
+            batch: Mini-batch of images.
```

Comment:
In the `training_step` "batch" refers to the output of the dataloader and in `forward` the argument batch just refers to the input images to the model as it is a tensor. Maybe forward should have a different argument name to differentiate it?

Reply:
What name would you suggest? Image?

Reply:
Went with "x" to match what we put in and h/z that we output.